### DISCUSSION TITLE
How to get the *full* custom-language UX in Notebooks?

### DISCUSSION THREAD

**Discussion by metaleap** [microsoft/vscode-discussions#1251] [Extension Development QnA] [answered]
_**TL;DR: I'm not seeing input-cell hovers / completions / etc for our functioning custom-language IntelliSense impl, nor output-cell syntax-coloring with `text/x-mylanguageid`, what am I missing?**_

___

So my [VSCode extension](https://github.com/metaleap/vscode-gerbil-scheme) based on `vscode-languageclient` and a custom backing [LSP](https://github.com/metaleap/gerbil-lsp/tree/main/lsp/msgs) **has all the essentials functioning well**: completions, hovers, document-symbols (and workspace symbols search), go-to-definition and list-references.

For our REPL UX, I'm doing a [custom NotebookSerializer and NotebookController](https://github.com/metaleap/vscode-gerbil-scheme/blob/main/src/repl.ts). Hooked up properly to [the `notebooks` contribution](https://github.com/metaleap/vscode-gerbil-scheme/blob/089834de68775c204bfdbc4b6846e0a3ea2886bb/package.json#L106C7-L114C8). They do work fine with regards to _serializing_ and _executing_. **BUT**, where's all that custom-language-extension UX goodness that I got working so perfectly well in the text-editor tabs?

### No completions, hover, go2def etc

In these Notebooks, other than syntax highlighting (for inputs, not outputs), there's **none** of all that well-working language-extension stuff that's already working in normal text editors for our language. For the input cell: no completions, no hover, no go2def via ctrl+click or F12, nothing! Just syntax coloring.

### No output-cell syntax coloring with `text/x-mylanguage`

With the cell output `mime`d to `text/x-gerbil` (the `gerbil` language ID [being properly registered](https://github.com/metaleap/vscode-gerbil-scheme/blob/089834de68775c204bfdbc4b6846e0a3ea2886bb/package.json#L52)), that output cell isn't syntax-highlighted at all (right now, the output is returned to be the input-cell's value so it's easy to tell: there is no highlighting). But changing it in my NotebookController code to `text/x-javascript` or `text/x-clojure` (VSCode builtins) **does** syntax-color the output cell! 

Here it is with `text/x-gerbil` output-cell `mime`, while the input cell has language ID `gerbil`:

![image](https://github.com/microsoft/vscode-discussions/assets/22109/ba88b5c5-d7a8-406f-8c61-7680712edf4d)

How come this fully-properly-registered custom language ID is seemingly second-class in Notebooks, when it's first-class in the normal code text-editor tabs? 

Am I missing some boilerplate that's not mentioned in VSCode's [Notebooks extension guide](https://code.visualstudio.com/api/extension-guides/notebook)?

**Comment by rebornix** [comment, marked as answer]
@metaleap thanks for your interest in notebook in VS Code! Our documentation is not up to date, will update them later but first let me answer some of your questions

Regarding language features, the notebook cell documents have an unique scheme `vscode-notebook-cell`, and in your language registration, it's limited to `file` https://github.com/metaleap/vscode-gerbil-scheme/blob/main/src/lsp.ts#L25C13-L26C1 , that's why none of them works for notebook. You can try add `{ language: 'gerbil', scheme: 'vscode-notebook-cell' }` to the document selector.

For "output-cell syntax coloring", we dropped support for builtin editor output and it's now fully contributed via output renderer https://github.com/rebornix/vscode-code-renderer , if you want to depend on this renderer, we could look into supporting your custom language in this renderer.

**Reply by metaleap** [reply]
> You can try add `{ language: 'gerbil', scheme: 'vscode-notebook-cell' }` to the document selector.

That did the trick! Many thanks @rebornix! :tada: :rocket: 

> Our documentation is not up to date

> we dropped support for builtin editor output and it's now fully contributed via output renderer

It's kinda odd though that it just-works when using `text/x-clojure` but not `text/x-gerbil`. Does that mean language-grammars from VSCode-owned "internal" extensions are internally "blessed" for output syntax coloring and other grammars are not? Wouldn't it anyway be the same grammar-to-color code, no matter whether the grammar was built-in or extension-provided?

> https://github.com/rebornix/vscode-code-renderer , if you want to depend on this renderer,

Not quite grokking this — is that repo of yours a dependency that current VSCode has, or is it some sort of 2.5-years-old "manual fork / extraction" of something buried also in VSCode's own repo?  =)

**Reply by rebornix** [reply]
> It's kinda odd though that it just-works when using text/x-clojure but not text/x-gerbil

@metaleap you got me. We have an internal renderer https://github.com/microsoft/vscode/blob/6c907814bfc223e842aa0d82e69a467ce4f76223/extensions/markdown-language-features/package.json#L40-L119 which lists all supported languages (the languages VS Code support). It currently doesn't support languages from external extensions but we could explore relaxing it and supporting rendering mime types which have language grammar provided by extensions.

**Reply by metaleap** [reply]
> but we could explore relaxing it and supporting rendering mime types which have language grammar provided by extensions

That'd be swell, here's hoping [the feature proposal issue I just made for this](https://github.com/microsoft/vscode/issues/223460) won't drown among the 8k+ other issues  :grin:

**Reply by faubulous** [reply]
@metaleap The documentation still stays 'text/x-LANGUAGE_ID for any other built-in or installed languages' which is wrong if I understand you correctly? Would be great if that could be corrected.

https://code.visualstudio.com/api/extension-guides/notebook#output-types
